### PR TITLE
Support more non-String Map keys of obvious dart:core types

### PR DESCRIPTION
**Pull request by kevmoo**
Partially addresses https://github.com/dart-lang/json_serializable/issues/396

### CODE REVIEW

File: json_serializable/lib/src/type_helpers/map_helper.dart [hunk 1]
```diff
@@ -56,9 +59,9 @@ class MapHelper extends TypeHelper<TypeHelperContextWithConfig> {
     _checkSafeKeyType(expression, keyArg);
 
     final valueArgIsAny = _isObjectOrDynamic(valueArg);
-    final isEnumKey = isEnum(keyArg);
+    final isKeyStringable = _isStringKeyable(keyArg);
```
Comment (natebosch) [inline json_serializable/lib/src/type_helpers/map_helper.dart]: why are the words flipped? What does Stringable mean? What does Keyable mean?

File: json_serializable/lib/src/type_helpers/map_helper.dart [hunk 2]
```diff
@@ -90,30 +93,58 @@ class MapHelper extends TypeHelper<TypeHelperContextWithConfig> {
...
+    }
+
     return '($expression $mapCast)$optionalQuestion.map('
         '($_keyParam, $closureArg) => MapEntry($keyUsage, $itemSubVal),)';
   }
 }
 
+final _intString = ToFromStringHelper('int.parse', 'toString()', 'int');
+
+final _instances = {
```
Comment (natebosch) [inline json_serializable/lib/src/type_helpers/map_helper.dart]: doc comment?

File: json_serializable/lib/src/type_helpers/map_helper.dart [hunk 3]
```diff
@@ -90,30 +93,58 @@ class MapHelper extends TypeHelper<TypeHelperContextWithConfig> {
...
+  bigIntString,
+  dateTimeString,
+  _intString,
+  uriString,
+};
+
+Iterable<String> get _allowedTypeNames => const [
```
Comment (natebosch) [inline json_serializable/lib/src/type_helpers/map_helper.dart]: allowed where? Doc comment?

File: json_serializable/lib/src/type_helpers/map_helper.dart [hunk 4]
```diff
@@ -90,30 +93,58 @@ class MapHelper extends TypeHelper<TypeHelperContextWithConfig> {
...
+      'Object',
+      'dynamic',
```
Comment (natebosch) [inline json_serializable/lib/src/type_helpers/map_helper.dart]: What is the implication of having `dynamic` here?

File: json_serializable/lib/src/type_helpers/map_helper.dart [hunk 5]
```diff
@@ -90,30 +93,58 @@ class MapHelper extends TypeHelper<TypeHelperContextWithConfig> {
...
+      'enum',
+      'String'
+    ].followedBy(_instances.map((i) => i.coreTypeName));
```
Comment (natebosch) [inline json_serializable/lib/src/type_helpers/map_helper.dart]: How about we bump the SDK constraint to 2.3.0 so we can do:

 ```dart
=> [
  'Object',
  'dynamic',
  'enum',
  'String',
  for ( var i in _instances )
    i.coreTypeName,
];
```

File: json_serializable/lib/src/type_helpers/map_helper.dart (hunk 1 above)
Comment (kevmoo) [reply json_serializable/lib/src/type_helpers/map_helper.dart]: Fix the function name. Added docs there.

File: json_serializable/lib/src/type_helpers/map_helper.dart (hunk 2 above)
Comment (kevmoo) [reply json_serializable/lib/src/type_helpers/map_helper.dart]: ack

File: json_serializable/lib/src/type_helpers/map_helper.dart (hunk 3 above)
Comment (kevmoo) [reply json_serializable/lib/src/type_helpers/map_helper.dart]: ack

File: json_serializable/lib/src/type_helpers/map_helper.dart (hunk 4 above)
Comment (kevmoo) [reply json_serializable/lib/src/type_helpers/map_helper.dart]: I can have a field of type `Map<dynamic, ...>` and things will "just work". The incoming JSON map (which, of course, has a `String` key) will be used directly.

File: json_serializable/lib/src/type_helpers/map_helper.dart (hunk 5 above)
Comment (kevmoo) [reply json_serializable/lib/src/type_helpers/map_helper.dart]: 🤷‍♂ many other places for that. No urgency here.

**Comment by kevmoo** [comment]
@natebosch – PTAL

**Pull request merged**